includes/class-wc-structured-data.php

```diff
@@ -0,0 +1,202 @@
+
+    foreach ( $this->data as $value ) {
+      $type = isset( $value['@type'] ) ? $value['@type'] : false;
+        
+      if ( 'Product' === $type || 'SoftwareApplication' === $type || 'MusicAlbum' === $type ) {
+        $products[] = $value;
+        $product_count ++;
+      } elseif ( 'Review' === $type ) {
+        $reviews[] = $value;
+      }
+    }
+    
+    if ( $product_count === 1 ) {
+      $structured_data = isset( $reviews ) ? $products[0] + array( 'review' => $reviews ) : $products[0];
+    } elseif ( $product_count > 1 ) {
+      $structured_data = array( '@graph' => $products );
+    }
+    
+    if ( ! isset( $structured_data ) ) {
+      return array();
+    }
+    
+    $context['@context'] = 'http://schema.org/';
+
+    return $context + $structured_data;
+  }
+
+  /**
+   * Contructor
+   */
+  public function __construct() {
+    add_action( 'woocommerce_before_shop_loop_item', array( $this, 'generate_product_category_data' ) );
+    add_action( 'woocommerce_single_product_summary', array( $this, 'generate_product_data' ) );
+    add_action( 'woocommerce_review_meta', array( $this, 'generate_product_review_data' ) );
+    add_action( 'wp_footer', array( $this, 'enqueue_data' ) );
+  }
+
+  /**
+   * Sanitizes, encodes and echoes the structured data into `wp_footer` action hook.
+   */
+  public function enqueue_data() {
+    if ( $structured_data = $this->get_data() ) {
+
+      array_walk_recursive( $structured_data, array( $this, 'sanitize_data' ) );
+
+      echo '<script type="application/ld+json">' . wp_json_encode( $structured_data ) . '</script>';
+    }
+  }
+
+  /**
+   * Callback function for sanitizing the structured data.
+   * Can't use anonymous function due to PHP v5.2 incompatibility...
+   *
+   * @param ref
+   */
+  private function sanitize_data( &$value ) {
+    $value = sanitize_text_field( $value );
+  }
+
+  /**
+   * Generates the product category structured data...
+   * Hooked into the `woocommerce_before_shop_loop_item` action hook...
+   */
+  public function generate_product_category_data() {
+    if ( ! is_product_category() ) {
+      return;
+    }
+    
+    $this->generate_product_data();
+  }
+  
+  /**
+   * Generates the product structured data...
+   * Hooked into the `woocommerce_single_product_summary` action hook...
+   * Applies the `woocommerce_product_structured_data` filter hook for clean structured data customization...
+   */
+  public function generate_product_data() {
+    global $product;
+
+    if ( $product->is_downloadable() ) {
+      switch ( $product->download_type ) {
+        case 'application' :
+          $type = "SoftwareApplication";
+        break;
+        case 'music' :
+          $type = "MusicAlbum";
+        break;
+        default :
+          $type = "Product";
+        break;
+      }
+    } else {
+      $type = "Product";
+    }
+
+    $json['@type']             = $type;
+    $json['@id']               = get_the_permalink();
```

Comment:
@leewillis77, I'm not sure here... Could you confirm this should be the permalink ?